### ISSUE TITLE
[.Net SDK, QnA Maker, Rich Cards] Code issues when build the QnA Bot with Hero Card

### ISSUE THREAD

**Issue by Jessica1980** [Microsoft/BotBuilder#4781]
## Bot Info
* SDK Platform: .NET
* SDK Version: 
* Active Channels: MS Teams, Direct Line
* Deployment Environment: Azure Bot Service

## Issue Description
I've made the QnA Bot in Azure. Then using this [video](https://www.c-sharpcorner.com/article/how-to-add-image-or-video-response-to-microsoft-azure-bot-application/) I've edited the code in Online App Service Editor. Then I went to Consol and made build.cmd. And received a number of issues regarding the new code.

## Code Example
`

	using System;
	using System.Threading;
	using System.Threading.Tasks;
	using Microsoft.Bot.Builder.Azure;
	using Microsoft.Bot.Builder.Dialogs;
	using Microsoft.Bot.Builder.CognitiveServices.QnAMaker;
	using Microsoft.Bot.Connector;
	
	namespace Microsoft.Bot.Sample.QnABot
	{
	    [Serializable]
	public class RootDialog : QnAMakerDialog
	//public class RootDialog : IDialog<object>
	{
	public async Task StartAsync(IDialogContext context)
	{
		/* Wait until the first message is received from the conversation and call MessageReceviedAsync 
		*  to process that message. */
		context.Wait(this.MessageReceivedAsync);
	}

	public RootDialog() : base(new QnAMakerService(new QnAMakerAttribute(ConfigurationManager.AppSettings[QnASubscriptionKey], ConfigurationManager.AppSettings[QnAKnowledgebaseId], "I don't know the answer", 0.5)))
	{
	}

	protected override async Task RespondFromQnAMakerResultAsync(IDialogContext context, IMessageActivity message, QnAMakerResults result)
	{
		Var answer = result.Answer.First().Answer;
		Activity reply = ((Activity)context.Activity).CreateReply();
		String[] qnaAnswerData = answer.Split(';');
		Int dataSize = qnaAnswerData.Length;
		// image and video card
		If (dataSize > 1 && dataSize <= 6)
		{
			Var attachment = GetSelectedCard(answer);
			Reply.Attachments.Add(attachment);
			Await context.PostAsync(reply);
		}
		Else
		{
			Await cantext.Forward(new BasicQnAMakerDialog(), AfterAnswerAsync, message, CancellationToken.None);
			Await context.PostAsync(reply);
		}
	}

	Private static Attachment GetSelectedCard(string answer)
	{
		Int len = answer.Split(';').Length;
		Switch (len)
		{
			Case 4: return GetHeroCard(answer);
			Case 6: return GetVideoCard(answer);
			Default: return GetHeroCard(answer);
		}
	}

	//Hero Card or Image Card
	Private static Attachment GetHeroCard(string answer)
	{
		String[] qnaAnswerData = answer.Split(';');
		String title = qnaAnswerData[0];
		String description = qnaAnswerData[1];
		String url = qnaAnswerData[2];
		String imageURL = qnaAnswerData[3];
		HeroCard card = new HeroCard
		{
			Title = title,
			Subtitle = description,
		};
		Card.Buttons = new List<CardAction>
		{
			New CardAction(ActionTypes.OpenUrl, "Learn More", value: url)
		};
		Card.Images = new List<CardImage>
		{
			New CardImage( url = imageURL)
		};
		Return card.ToAttachment();
	}

	//Video Card
	Private static Attachment GetVideoCard(string answer)
	{
		String[] qnaAnswerData = answer.Split(';');
		String title = qnaAnswerData[0];
		String subtitle = qnaAnswerData[1];
		String description = qnaAnswerData[2];
		String thumbimageurl = qnaAnswerData[3];
		String mediaURL = qnaAnswerData[4];
		String URL = qnaAnswerData[5];
		VideoCard card = new VideoCard
		{
			Title = title,
			Subtitle = subtitle,
			Text = description,
		};
		Card.Image = new ThumbnailUrl
		{
			Url = thumbimageurl
		};
		Card.Media = new List<MediaUrl>
		{
			New MediaUrl()
			{
				Url = mediaUrl
			}
		};
		Card.Buttons = new List<CardAction>
		{
			New CardAction()
			{
				Title= "Learn More",
				Type = ActionTypes.OpenUrl,
				Value = url
			}
		};
		Return card.ToAttachment();
	}`

## Expected Behavior
Build without issues

## Actual Results
> Handling .NET Web Application deployment.
MSBuild auto-detection: using msbuild version '14.0' from 'D:\Program Files (x86)\MSBuild\14.0\bin\amd64'.
All packages listed in packages.config are already installed.
Dialogs\BasicQnAMakerDialog.cs(34,37): error CS1002: ; expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(38,17): error CS1003: Syntax error, ',' expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(38,18): error CS1002: ; expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(40,7): error CS1002: ; expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(42,17): error CS1003: Syntax error, ',' expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(42,18): error CS1002: ; expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(43,17): error CS1003: Syntax error, ',' expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(43,18): error CS1002: ; expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(47,10): error CS1585: Member modifier 'static' must precede the member type and name [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(50,15): error CS1002: ; expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(52,9): error CS1002: ; expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(52,10): error CS1002: ; expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(52,10): error CS1513: } expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(53,9): error CS1002: ; expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(53,10): error CS1002: ; expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(53,10): error CS1513: } expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(59,10): error CS1585: Member modifier 'static' must precede the member type and name [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(73,8): error CS1003: Syntax error, ',' expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(77,8): error CS1003: Syntax error, ',' expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(79,14): error CS1003: Syntax error, ',' expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(79,15): error CS1002: ; expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(83,10): error CS1585: Member modifier 'static' must precede the member type and name [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(104,8): error CS1003: Syntax error, ',' expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(104,18): error CS1003: Syntax error, ',' expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(111,8): error CS1003: Syntax error, ',' expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(111,20): error CS1003: Syntax error, ',' expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(118,14): error CS1003: Syntax error, ',' expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(118,15): error CS1002: ; expected [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Failed exitCode=1, command="D:\Program Files (x86)\MSBuild\14.0\Bin\MSBuild.exe" "D:\home\site\wwwroot\.\Microsoft.Bot.Sample.QnABot.csproj" /nologo /verbosity:m /t:Build /p:AutoParameterizationWebConfigConnectionStrings=false;Configuration=Release;UseSharedCompilation=false /p:SolutionDir="D:\home\site\wwwroot\.\.\\"
An error has occurred during web site deployment.

**Comment by Jessica1980** [comment]
@Kartik1607 thank you for the help! I've checked the code and deleted mistakes.

Now I've received only 2 issues, but cann't understand what I should do :( Looks like the video is about thr previous QnA Maker and I have the base in the previous QnA Maker.

> Dialogs\BasicQnAMakerDialog.cs(27,39): error CS0115: 'RootDialog.RespondFormQnAMakerResultAsync(IDialogContext, IMessageActivity, QnAMakerResults)': no suitable method found to override [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(126,28): warning CS0108: 'RootDialog.MessageReceivedAsync(IDialogContext, IAwaitable<IMessageActivity>)' hides inherited member 'QnAMakerDialog.MessageReceivedAsync(IDialogContext, IAwaitable<IMessageActivity>)'. Use the new keyword if hiding was intended. [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Failed exitCode=1, command="D:\Program Files (x86)\MSBuild\14.0\Bin\MSBuild.exe" "D:\home\site\wwwroot\.\Microsoft.Bot.Sample.QnABot.csproj" /nologo /verbosity:m /t:Build /p:AutoParameterizationWebConfigConnectionStrings=false;Configuration=Release;UseSharedCompilation=false /p:SolutionDir="D:\home\site\wwwroot\.\.\\"
An error has occurred during web site deployment.

Full code:
`
	using System;
	using System.Threading;
	using System.Threading.Tasks;
	using Microsoft.Bot.Builder.Azure;
	using Microsoft.Bot.Builder.Dialogs;
	using Microsoft.Bot.Builder.CognitiveServices.QnAMaker;
	using Microsoft.Bot.Connector;

namespace Microsoft.Bot.Sample.QnABot
{
    [Serializable]
    public class RootDialog : QnAMakerDialog
    {
        public async Task StartAsync(IDialogContext context)
        {
            /* Wait until the first message is received from the conversation and call MessageReceviedAsync 
            *  to process that message. */
            context.Wait(this.MessageReceivedAsync);
        }
        
        public RootDialog() : base(new QnAMakerService(new QnAMakerAttribute(ConfigurationManager.AppSettings["QnaSubscriptionKey"], ConfigurationManager.AppSettings["QnaKnowledgebaseId"], "No good match in FAQ.", 0.5)))
        {
            
        }
        
        protected override async Task RespondFormQnAMakerResultAsync(IDialogContext context, IMessageActivity message, QnAMakerResults result)
        {
            var answer = result.AnswersFirst().Answer;
		Activity reply = ((Activity)context.Activity).CreateReply();

		string[] qnaAnswerData = answer.Split(';');
		int dataSize = qnaAnswerData.Length;
		//image and video card
		if (dataSize > 1 && dataSize <= 6)
		{
			var attachment = GetSelectedCard(answer);
			reply.Attachments.Add(attachment);
			await context.PostAsync(reply);
		}
		else
		{
			await context.Forward(new BasicQnAMakerDialog(), AfterAnswerAsync, message, CancellationToken.None);
            await context.PostAsync(reply);
		}
        }
        
        private static Attachment GetSelectedCard(string answer)
	{
		int len = answer.Split(';').Length;
		switch (len)
		{
			case 4: return GetHeroCard(answer);
			case 6: return GetVideoCard(answer);
			default: return GetHeroCard(answer);
		}
	}
        
        //Hero Card
        private static Attachment GetHeroCard(string answer)
	{
		string[] qnaAnswerData = answer.Split(';');
		string title = qnaAnswerData[0];
		string description = qnaAnswerData[1];
		string url = qnaAnswerData[2];
		string imageURL = qnaAnswerData[3];
		HeroCard card = new HeroCard
		{
			Title = title,
			Subtitle = description,
		};

		card.Buttons = new List<CardAction>
		{
			new CardAction(ActionTypes.OpenUrl, "Learn More", value: url)
		};

		card.Images = new List<CardImage>
		{
			new CardImage( url = imageURL)
		};
		return card.ToAttachment();
	}
        
        //Video Card
        private static Attachment GetVideoCard(string answer)
	{
		string[] qnaAnswerData = answer.Split(';');
		string title = qnaAnswerData[0];
		string subtitle = qnaAnswerData[1];
		string description = qnaAnswerData[2];
		string thumbimageurl = qnaAnswerData[3];
		string mediaURL = qnaAnswerData[4];
		string url= qnaAnswerData[5];
		VideoCard card = new VideoCard
		{
			Title = title,
			Subtitle = subtitle,
			Text = description,
		};

		card.Image = new ThumbnailUrl
		{
			Url = thumbimageurl
		};
		card.Media = new List<MediaUrl>
			{
				new MediaUrl()
				{
					Url = mediaUrl
				}
			};		

		card.Buttons = new List<CardAction>
		{
			new CardAction()
			{
				Title = "Learn More",
				Type = ActionTypes.OpenUrl,
				Value = url
			}
		};
		return card.ToAttachment();
	}

        private async Task MessageReceivedAsync(IDialogContext context, IAwaitable<IMessageActivity> result)
        {
            /* When MessageReceivedAsync is called, it's passed an IAwaitable<IMessageActivity>. To get the message,
             *  await the result. */
            var message = await result;

            var qnaAuthKey = GetSetting("QnAAuthKey"); 
            var qnaKBId = Utils.GetAppSetting("QnAKnowledgebaseId");
            var endpointHostName = Utils.GetAppSetting("QnAEndpointHostName");

            // QnA Subscription Key and KnowledgeBase Id null verification
            if (!string.IsNullOrEmpty(qnaAuthKey) && !string.IsNullOrEmpty(qnaKBId))
            {
                // Forward to the appropriate Dialog based on whether the endpoint hostname is present
                if (string.IsNullOrEmpty(endpointHostName))
                    await context.Forward(new BasicQnAMakerPreviewDialog(), AfterAnswerAsync, message, CancellationToken.None);
                else
                    await context.Forward(new BasicQnAMakerDialog(), AfterAnswerAsync, message, CancellationToken.None);
            }
            else
            {
                await context.PostAsync("Please set QnAKnowledgebaseId, QnAAuthKey and QnAEndpointHostName (if applicable) in App Settings. Learn how to get them at https://aka.ms/qnaabssetup.");
            }

        }

        private async Task AfterAnswerAsync(IDialogContext context, IAwaitable<IMessageActivity> result)
        {
            // wait for the next user message
            context.Wait(MessageReceivedAsync);
        }

        public static string GetSetting(string key)
        {
            var value = Utils.GetAppSetting(key);
            if (String.IsNullOrEmpty(value) && key == "QnAAuthKey")
            {
                value = Utils.GetAppSetting("QnASubscriptionKey"); // QnASubscriptionKey for backward compatibility with QnAMaker (Preview)
            }
            return value;
        }
    }

    // Dialog for QnAMaker Preview service
    [Serializable]
    public class BasicQnAMakerPreviewDialog : QnAMakerDialog
    {
        // Go to https://qnamaker.ai and feed data, train & publish your QnA Knowledgebase.
        // Parameters to QnAMakerService are:
        // Required: subscriptionKey, knowledgebaseId, 
        // Optional: defaultMessage, scoreThreshold[Range 0.0 – 1.0]
        public BasicQnAMakerPreviewDialog() : base(new QnAMakerService(new QnAMakerAttribute(RootDialog.GetSetting("QnAAuthKey"), Utils.GetAppSetting("QnAKnowledgebaseId"), "No good match in FAQ.", 0.5)))
        { }
    }

    // Dialog for QnAMaker GA service
    [Serializable]
    public class BasicQnAMakerDialog : QnAMakerDialog
    {
        // Go to https://qnamaker.ai and feed data, train & publish your QnA Knowledgebase.
        // Parameters to QnAMakerService are:
        // Required: qnaAuthKey, knowledgebaseId, endpointHostName
        // Optional: defaultMessage, scoreThreshold[Range 0.0 – 1.0]
        public BasicQnAMakerDialog() : base(new QnAMakerService(new QnAMakerAttribute(RootDialog.GetSetting("QnAAuthKey"), Utils.GetAppSetting("QnAKnowledgebaseId"), "No good match in FAQ.", 0.5, 1, Utils.GetAppSetting("QnAEndpointHostName"))))
        { }

    }
}`

**Comment by Kartik1607** [comment]
Seems like you are missing imports.

**Comment by Jessica1980** [comment]
@Kartik1607 what "import" is? I use online code editor in Azure, after I put all changes in the code I go to console and write build.cmd command.
Or do you mean something different?

**Comment by Jessica1980** [comment]
I've added "using System.Configuration;" - this help to resolve two issues CS0103.
But how to resolve issues CS1061, CS0246 and CS0108 still have no idea :(
Could you please help?

**Comment by Jessica1980** [comment]
The issues CS0246 have gone after I've added "using System.Collections.Generic;".

So I have now only 2 issues - CS1061 and CS0246. How can I resolve it?

> Handling

 .NET Web Application deployment.
MSBuild auto-detection: using msbuild version '14.0' from 'D:\Program Files (x86)\MSBuild\14.0\bin\amd64'.
All packages listed in packages.config are already installed.
Dialogs\BasicQnAMakerDialog.cs(132,28): warning CS0108: 'RootDialog.MessageReceivedAsync(IDialogContext, IAwaitable<IMessageActivity>)' hides inherited member 'QnAMakerDialog.MessageReceivedAsync(IDialogContext, IAwaitable<IMessageActivity>)'. Use the new keyword if hiding was intended. [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Dialogs\BasicQnAMakerDialog.cs(32,31): error CS1061: 'List<QnAMakerResult>' does not contain a definition for 'First' and no extension method 'First' accepting a first argument of type 'List<QnAMakerResult>' could be found (are you missing a using directive or an assembly reference?) [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Failed exitCode=1, command="D:\Program Files (x86)\MSBuild\14.0\Bin\MSBuild.exe" "D:\home\site\wwwroot\.\Microsoft.Bot.Sample.QnABot.csproj" /nologo /verbosity:m /t:Build /p:AutoParameterizationWebConfigConnectionStrings=false;Configuration=Release;UseSharedCompilation=false /p:SolutionDir="D:\home\site\wwwroot\.\.\\"
An error has occurred during web site deployment.

**Comment by Jessica1980** [comment]
And I've put the part on 132 string in /*...*/ - so the issue CS0108 is resolved. And I still have only one issue - CS 1061 and still need your help for resolve it:

> Handling .NET Web Application deployment.
MSBuild auto-detection: using msbuild version '14.0' from 'D:\Program Files (x86)\MSBuild\14.0\bin\amd64'.
All packages listed in packages.config are already installed.
Dialogs\BasicQnAMakerDialog.cs(32,31): error CS1061: 'List<QnAMakerResult>' does not contain a definition for 'First' and no extension method 'First' accepting a first argument of type 'List<QnAMakerResult>' could be found (are you missing a using directive or an assembly reference?) [D:\home\site\wwwroot\Microsoft.Bot.Sample.QnABot.csproj]
Failed exitCode=1, command="D:\Program Files (x86)\MSBuild\14.0\Bin\MSBuild.exe" "D:\home\site\wwwroot\.\Microsoft.Bot.Sample.QnABot.csproj" /nologo /verbosity:m /t:Build /p:AutoParameterizationWebConfigConnectionStrings=false;Configuration=Release;UseSharedCompilation=false /p:SolutionDir="D:\home\site\wwwroot\.\.\\"
An error has occurred during web site deployment.

**Comment by Kartik1607** [comment]
Seems you are new to coding in general. I'll help you in learning how to read error.
The error is 

`Dialogs\BasicQnAMakerDialog.cs(32,31): error CS1061: 'List' does not contain a definition for 'First' and no extension method 'First' accepting a first argument of type 'List' could be found (are you missing a using directive or an assembly reference?) `

`Dialogs\BasicQnAMakerDialog.cs(32,31):` tells us that the error is in `BasicQnAMakerDialog.cs` file at line number `32`.

Then, we need to figure out the error.  'List' **does not contain a definition** for **'First'** and no extension method **'First' accepting a first argument of type 'List'** could be found

From this information, I can infer that on line 32, you are using a List element and calling the **First** method, but you are also passing in an argument in the function. The argument is also of type List.

Now, if you look at documentation for List at [MSDN](https://msdn.microsoft.com/en-us/library/6sh2ey19(v=vs.110).aspx) and search for `First`, you will find all the possible arguments you can send in the method.

**Comment by Jessica1980** [comment]
@Kartik1607 thank you for your helping! You are right - I'm a newby in codding. I decided to start learning with the templates and want to creat the QnA Bot with Hero Card and thought that I can do it with the vifeo instruction... But it isn't easy as I thought :)
I checked the line 32 - I do not use a List. I use List 76 line:
![1](https://user-images.githubusercontent.com/40513670/41819764-c0c66042-77ce-11e8-8c54-be97bf6adb1c.png)
![2](https://user-images.githubusercontent.com/40513670/41819765-c0e15f6e-77ce-11e8-97bc-2111910b9543.png)

I'm reading MSDN (thank you again for the link!), but it is dificult for me (at my current level of coding) to understand what I did wrong.. In the video insruction was this code that I show on screenshots.
Can I ask you to help me to understand where I'm wrong?
